applications/serial_lte_modem/src/mqtt_c/slm_at_mqtt.c

```diff
@@ -217,41 +217,70 @@ static void mqtt_thread_fn(void *arg1, void *arg2, void *arg3)
 	while (true) {
 		if (!ctx.connected) {
 			LOG_WRN("MQTT disconnected");
+			err = 0;
 			break;
 		}
 		err = poll(&fds, 1, mqtt_keepalive_time_left(&client));
 		if (err < 0) {
 			LOG_ERR("ERROR: poll %d", errno);
 			break;
 		}
-		/* timeout or revent, send KEEPALIVE */
-		(void)mqtt_live(&client);
 
 		if ((fds.revents & POLLIN) == POLLIN) {
 			err = mqtt_input(&client);
 			if (err != 0) {
 				LOG_ERR("ERROR: mqtt_input %d", err);
-				mqtt_abort(&client);
+				break;
+			}
+			/* MQTT v3.1.1: If a Client does not receive a PINGRESP Packet within a
+			 *  reasonable amount of time after it has sent a PINGREQ, it SHOULD close
+			 *  the Network Connection to the Server.
+			 */
+			if (client.unacked_ping > 1) {
+				LOG_ERR("ERROR: mqtt_ping nack %d", client.unacked_ping);
+				err = -ENETRESET;
 				break;
 			}
 		}
+
+		/* MQTT v3.1.1: Note that a Server is permitted to disconnect a Client that it
+		 * determines to be inactive or non-responsive at any time, regardless of the
+		 * Keep Alive value provided by that Client.
+		 */
 		if ((fds.revents & POLLERR) == POLLERR) {
 			LOG_ERR("POLLERR");
-			mqtt_abort(&client);
+			err = -EIO;
 			break;
 		}
 		if ((fds.revents & POLLHUP) == POLLHUP) {
 			LOG_ERR("POLLHUP");
-			mqtt_abort(&client);
+			err = -ECONNRESET;
 			break;
 		}
 		if ((fds.revents & POLLNVAL) == POLLNVAL) {
 			LOG_ERR("POLLNVAL");
-			mqtt_abort(&client);
+			err = -ENOTCONN;
+			break;
+		}
+
+		/* poll timeout or revent, send KEEPALIVE */
+		err = mqtt_live(&client);
+		if (err != 0 && err != -EAGAIN) {
+			LOG_ERR("ERROR: mqtt_live %d", err);
 			break;
 		}
 	}
 
+	if (ctx.connected && err != 0) {
```

Comment:
Shouldn't this terminate the connection if ctx.connected is true at this point? It would be simpler and would allow the POLLHUP to return 0.

Reply:
This is to protect `#XMQTTCON=0`, without it there will be two URC, `: 1,0`  and `: 1,-113`